File: ice0208/binary-tree-dfs/872.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def go_down(self, cursor, leaf_list: list[int]):
        if not cursor.left and not cursor.right:
            leaf_list.append(cursor.val)
            return

        if cursor.left:
            self.go_down(cursor.left, leaf_list)
        if cursor.right:
            self.go_down(cursor.right, leaf_list)

    def leafSimilar(self, root1, root2) -> bool:
        leaf1, leaf2 = [], []

        if root1:
            self.go_down(root1, leaf1)
        if root2:
            self.go_down(root2, leaf2)

        return leaf1 == leaf2

File: moonpoison/binary_tree-dfs/872.cpp
#include<bits/stdc++.h>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution {
public:
    vector<int> b;
    bool leafSimilar(TreeNode* root1, TreeNode* root2) 
    {
        dfs(root1);
        vector<int> a=b;
        b.clear();
        dfs(root2);
        return a==b;
    }
    void dfs(TreeNode* root)
    {
        if(root->left!=nullptr) dfs(root->left);
        if(root->right!=nullptr) dfs(root->right);
        if(root->left==nullptr&&root->right==nullptr) b.push_back(root->val);
    }
};